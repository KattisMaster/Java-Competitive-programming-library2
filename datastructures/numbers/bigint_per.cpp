/* KTH ACM Contest Template Library
 *
 * Numeric/Unsigned Big Integer/Simple
 *
 * Features:
 * # read/write in base 10 from/to i-/ostream
 * # cmp/add/sub
 * # multiplication [dependancy: addition]
 * # division and modulo [dependancy for arbitrary denominator: add, sub]
 * # exponentiation [dependancy: multiplication]
 * # find n:th root of (SLOW) [dependancy: exp., division, comparison]
 * # grows dynamically when In Need.
 *
 * What's bad?
 * # not bit-oriented, therefore no bit ops
 *   (e.g. and, not, shift, etc)
 * # no I/O in other bases (easy to implement naively though)
 * # far from the Fastest Thing Possible (TM).
 *
 *
 * Credit:
 *   By
 *        Per Austrin, pego@kth.se 2002-09-15
 *   Updated by
 *        Per Austrin, Christer Stålstrand, 2002-09-26
 *   Bugfixes (fix in modulo and input/output)
 *        Per Austrin, 2002-10-05
 *   Bugfixes and update (fixes in add and sub, code compression)
 *        Per Austrin, 2003-03-14
 *   Update (the long called-for divmod with arbitrary denominator!!
 *           and more code compression of course)
 *        Per Austrin, 2003-09-22
 *
 */

#include <iostream>
#include <iomanip>
#include <string>
#include <vector>

/* if long longs are disallowed, use the following alternative
 * definitions (the rest should work by magic):
 * #define LSIZE 10000
 * #define LIMBDIGS 4
 * typedef int limb;     */

#define LSIZE 1000000000 /* 10^9 */
#define LIMBDIGS 9

/* in order for subtraction to work properly, limb needs to be signed. */
typedef long long limb;
typedef vector<limb> bigint;
typedef bigint::const_iterator bcit;
typedef bigint::reverse_iterator brit;
typedef bigint::const_reverse_iterator bcrit;
typedef bigint::iterator bit;


bigint BigInt(limb i) {
  bigint res;
  do res.push_back(i % LSIZE);
  while (i /= LSIZE);
  return res;
}

istream& operator>>(istream& i, bigint& n) {
  string s;
  i >> s;
  int l = s.length();
  n.clear();
  while (l > 0) {
    int j = 0;
    for (int k = l > LIMBDIGS ? l-LIMBDIGS: 0; k < l; ++k)
      j = 10*j + s[k]-'0';
    n.push_back(j);
    l -= LIMBDIGS;
  }
  return i;
}

/* Warning: the ostream must be configured to print things with right
 * justification, lest output be ooky.    */
ostream& operator<<(ostream& o, const bigint& n) {
  int began = 0, ofill = o.fill();
  o.fill('0');
  for (bcrit i = n.rbegin(); i != n.rend(); ++i) {
    if (began) o << setw(LIMBDIGS);
    if (*i) began = 1;
    if (began) o << *i;
  }
  if (!began) o << "0";
  o.fill(ofill);
  return o;
}

/* The base comparison function. semantics like strcmp(...).   */
int cmp(const bigint& n1, const bigint& n2) {
  int x = n2.size() - n1.size();
  bcit i = n1.end() - 1, j = n2.end() - 1;
  while (x-- > 0) if (*j--) return -1;
  while (++x < 0) if (*i--) return 1;
  // above replace below (untested)
  /*
  if (x > 0) {
    while (x--)
      if (*j--) return -1;
  } else if (x < 0) {
    while (x++)
      if (*i--) return 1;
  }
  */
  for (; i + 1 != n1.begin(); --i, --j)
    if (*i != *j)
      return *i-*j;
  return 0;
}

/* The other operators will be automatically defined by STL */
bool operator==(const bigint& n1, const bigint& n2) { return !cmp(n1,n2); }
bool operator<(const bigint& n1, const bigint& n2) { return cmp(n1,n2) < 0; }

bigint& operator+=(bigint& a, const bigint& b) {
  if (a.size() < b.size()) a.resize(b.size());
  limb cy = 0;
  bit i = a.begin();
  for (bcit j = b.begin(); i != a.end() && (cy || j < b.end()); ++j, ++i)
    cy += *i + (j < b.end() ? *j : 0), *i = cy % LSIZE, cy /= LSIZE;
  if (cy) a.push_back(cy);
  return a;
}

/* Returns true if sign changed   */
bool sub(bigint& a, const bigint& b) {
  if (a.size() < b.size()) a.resize(b.size());
  limb cy = 0;
  bit i = a.begin();
  for (bcit j = b.begin(); i != a.end() && (cy || j < b.end()); ++j, ++i) {
    *i -= cy + (j < b.end() ? *j : 0);
    if ((cy = *i < 0)) *i += LSIZE;
  }
  /* If sign changed, flip all digits.  These three lines can be
   * ignored if it is known that the sign will not change, e.g. when
   * using bigint in conjunction with sign.cpp, in the divmod method,
   * or in many combinatorial counting problems.  */
  if (cy)
    while (i-- > a.begin())
      *i = LSIZE - *i;
  return cy;
}
bigint& operator-=(bigint& a, const bigint& b) { sub(a, b); return a; }

bigint& operator*=(bigint& a, limb b) {
  limb cy = 0;
  for (bit i = a.begin(); i != a.end(); ++i)
    cy = cy / LSIZE + *i * b, *i = cy % LSIZE;
  while (cy /= LSIZE) a.push_back(cy % LSIZE);
  return a;
}

bigint& operator*=(bigint& a, const bigint& b) {
  bigint x = a, y, bb = b;
  a.clear();
  for (bcit i = bb.begin(); i != bb.end(); ++i)
    (y = x) *= *i, a += y, x.insert(x.begin(), 0);
  return a;
}

/* a will hold floor(a/b), rest will hold a % b  */
bigint& divmod(bigint& a, limb b, limb* rest = NULL) {
  limb cy = 0;
  for (brit i = a.rbegin(); i != a.rend(); ++i)
    cy += *i, *i = cy / b, cy = (cy % b) * LSIZE;
  if (rest)
    *rest = cy / LSIZE;
  return a;
}

/* a will hold a % b (and is returned), quo will hold floor(a/b).
 * NB!! different semantics from one-limb divmod!!
 * NB!! quo should be different from a!! */
bigint& divmod(bigint& a, const bigint& b, bigint* quo = NULL) {
  /* If divmod is never called with b the same as a or quo
   * this copying is not necessary. */
  bigint den = b;
  brit j = den.rbegin(), i = a.rbegin();
  for ( ; j != den.rend() && !*j; ++j);
  for ( ; i != a.rend() && !*i; ++i);
  int n = a.rend() - i, m = den.rend() - j;
  if (!m) { /* Division by zero! */ abort(); }
  /* If divisor is limbsized, resort to regular linear-time divmod. */
  if (m == 1) {
    bigint q;
    return (quo ? *quo : q) = a, a.resize(1), 
      divmod(quo ? *quo : q, *j, &a.front()), a;
  }
  
  bigint tmp;
  /* Use first two digits for a good estimate of the quotient.
   * (Though this function could probably be a lot shorter if we just
   * used one digit instead)  */
  limb den0 = (*++j + *--j * LSIZE) + 1;
  if (quo) quo->clear();
  /* Loop invariant: quo * den + a = num */
  while (a >= den) {
    limb num0 = (*++i + *--i * LSIZE), z = num0 / den0, cy = 0;
    /* Silly degenerate case: */
    if (z == 0 && n == m) z = 1;
    tmp.resize(n - m - !z);
    /* Non-silly degenerate case: */
    if (!z) z = num0 / (*j + 1);
    /* Set tmp = z * b^(n-m) and add to quotient */
    if (quo) tmp.push_back(z), *quo += tmp, tmp.pop_back();
    /* Set tmp = den * z * b^(n-m)  */
    for (bcit j = den.begin(); j != den.end(); ++j)
      cy += *j * z, tmp.push_back(cy % LSIZE), cy /= LSIZE;
    if (cy) tmp.push_back(cy);
    /* Note that we rely on the important fact that tmp doesn't have
     * more limbs than a, so that a won't be resized, and the iterator
     * i will still valid. */
    if (tmp.size() > a.size()) tmp.resize(a.size());
    sub(a, tmp);
    while (i != a.rend() && !*i) --n, ++i;
  }
  return a; 
}


bigint& operator/=(bigint& a, const bigint& b) {
  bigint q;
  return divmod(a, b, &q), a = q;
}
bigint& operator%=(bigint& a, const bigint& b) { return divmod(a, b, NULL); }

bigint& operator/=(bigint& a, limb b) { return divmod(a, b); }
limb operator%(const bigint& a, limb b) {
  limb res;
  bigint fubar = a;
  return divmod(fubar, b, &res), res;
}


/* Auxilliary stuff which should probably be moved out.
 */ 


/* Exponentiation. */
bigint& operator^=(bigint& a, limb b) {
  bigint aa = a;
  a.clear(); a.push_back(1);
  while (b) {
    if (b & 1) a *= aa;
    aa *= aa;
    b >>= 1;
  }
  return a;
}

/* Finds the e:th root of n in far worse time than necessary.
 * Returns 0 if the root doesn't exist.    */
bigint root(const bigint& n, limb e) {
  int f;
  bigint lo = BigInt(0), hi, m, n2;
  hi = BigInt(LSIZE);
  // let hi ~ LS^(1+log_LS(n))/e) = O(n^(1 + 1/e))
  hi ^= ((n.size()+1) / e) + 1;
  while (1) {
    (m = lo) += hi;
    divmod(m, 2);
    if (m == lo) break;
    (n2 = m) ^= e;
    f = cmp(n, n2);
    if (f < 0) hi = m;
    else if (f > 0) lo = m;
    else return m;
  }
  /* If just an approximation of the root is wanted, change return
   * statement to:
   * return lo;    (for floor(root))
   * return hi;    (for ceil(root))          */
  return BigInt(0);
}

