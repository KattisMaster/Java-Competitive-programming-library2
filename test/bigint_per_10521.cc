/* @JUDGE_ID:    ..  10521   C++ */

#include "../datastructures/numbers/bigint_per.cpp"

const enum {SIMPLE, FOR, WHILE} mode = WHILE;
const bool debug = false;

void init() {}

const bigint zero;
const bigint one = BigInt(1);

void gen_frac(bigint& n, bigint& m, bool neg) {
  bigint q;
  divmod(n, m, &q);
  if (neg && n != zero) {
    bigint tmp = n;
    q += one;
    (n = m) -= tmp;
  }
  if (neg && q != zero) cout << '-';
  cout << q;
  if (n != zero) {
    cout << "+1/{";
    gen_frac(m, n, false);
    cout << "}";
  }
}

bool solve(int P) {
  bool neg = false;
  bigint n, m;

  while (cin && isspace(cin.peek())) cin.get();
  if (cin && cin.peek() == '-') cout << '-', neg = !neg, cin.get();
  if (!(cin >> n)) return false;
  cout << n << "/";

  while (cin && isspace(cin.peek())) cin.get();
  if (cin && cin.peek() == '-') cout << '-', neg = !neg, cin.get();
  if (!(cin >> m)) return false;
  cout << m << " = ";

  gen_frac(n, m, neg);
  cout << endl;
  return true;
}

int main() {
  init();
  int n = mode == SIMPLE ? 1 : 1<<30;
  if (mode == FOR) cin >> n;
  for (int i = 0; i < n && solve(i); ++i);
  return 0;
}
