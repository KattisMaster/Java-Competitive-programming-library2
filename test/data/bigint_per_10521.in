34 -64
9 -5
5 6
1 1
375283957328957832975298351364782365238763287462317864 235623756238756832756782365187463127856187567813
347589345734856743658723641328746178461 4621387462378463287462378462378462387463287463287563248756132785613287
2365728365238756238756178461785612387563478561785624986523974623897561378461387563187 562318756327895693257892365218563219875623398756
3748923572365723865832756328752656235823165278356238756238756237856238756 742389472389478932749832748932657893465783465872365873246587346587342658374568734
07156217803516290126375612712576239586129561928756283795623789562378915612793856219835623879567321856231578216352357238957238957238915621785689650726521789461239321751107501758291752365012657126357102521307865012786501256015601010120965320715621780351629012637561271257623958612956192875628379562378956237891561279385621983562387956732185623157821635 27589123758192057892137580219375890123758091237589016718297682197689127638496713896713489678134907608932147629837610987618920768591437589127512938758123957129857210957213895721389572139857213895721985712389571230985721389047123980472198347829137589127618927528913571328957129867129471209572198768901237509821375890213758219305789132075
0 1
0     3278
-214     2347826847632478236
19   1718
-23578 -273589
-123432521751 -23521512