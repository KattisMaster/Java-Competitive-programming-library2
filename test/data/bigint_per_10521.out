34/-64 = -1+1/{2+1/{7+1/{2}}}
9/-5 = -2+1/{5}
5/6 = 0+1/{1+1/{5}}
1/1 = 1
375283957328957832975298351364782365238763287462317864/235623756238756832756782365187463127856187567813 = 1592725+1/{2+1/{7+1/{4+1/{1+1/{2+1/{1+1/{1+1/{4+1/{15+1/{1+1/{18+1/{2+1/{1+1/{10+1/{4+1/{11+1/{5+1/{1+1/{2+1/{22+1/{1+1/{8+1/{1+1/{2+1/{1+1/{2+1/{3+1/{1+1/{42+1/{3+1/{1+1/{3+1/{9+1/{14+1/{7+1/{1+1/{6+1/{1+1/{3+1/{1+1/{11+1/{5+1/{4+1/{17+1/{9+1/{21+1/{2+1/{5+1/{2+1/{3+1/{1+1/{1+1/{1+1/{2+1/{3+1/{2+1/{1+1/{1+1/{2+1/{3+1/{1+1/{6+1/{1+1/{10+1/{1+1/{35+1/{2+1/{4+1/{5+1/{2+1/{2+1/{2+1/{2+1/{2+1/{4+1/{3+1/{1+1/{1+1/{1+1/{2+1/{1+1/{4+1/{1+1/{3+1/{1+1/{1+1/{8+1/{5+1/{1+1/{1+1/{1+1/{1+1/{11}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}
347589345734856743658723641328746178461/4621387462378463287462378462378462387463287463287563248756132785613287 = 0+1/{13295538310036941148926248792184+1/{1+1/{1+1/{6+1/{1+1/{23+1/{5+1/{2+1/{13+1/{16+1/{2+1/{3+1/{1+1/{1+1/{1+1/{6+1/{3+1/{2+1/{1+1/{1+1/{1+1/{9+1/{1+1/{2+1/{9+1/{2+1/{1+1/{2+1/{2+1/{11+1/{1+1/{1+1/{2+1/{4+1/{1+1/{1+1/{6+1/{1+1/{1+1/{4+1/{3+1/{1+1/{1+1/{5+1/{1+1/{1+1/{3+1/{4+1/{5+1/{6+1/{11+1/{1+1/{1+1/{2+1/{1+1/{4+1/{1+1/{1+1/{1+1/{3+1/{77+1/{1+1/{11+1/{3+1/{1+1/{1+1/{4+1/{1+1/{4+1/{2+1/{2+1/{54+1/{1+1/{1+1/{30+1/{1+1/{1+1/{2+1/{1+1/{1+1/{5+1/{8}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}
2365728365238756238756178461785612387563478561785624986523974623897561378461387563187/562318756327895693257892365218563219875623398756 = 4207094888115857091943940751175235978+1/{4+1/{14+1/{1+1/{1+1/{3+1/{6+1/{1+1/{4+1/{8+1/{1+1/{4+1/{3+1/{2+1/{2+1/{3+1/{1+1/{1+1/{2+1/{1+1/{732+1/{1+1/{17+1/{2+1/{2+1/{2+1/{3+1/{8+1/{2+1/{2+1/{213+1/{2+1/{3+1/{5+1/{1+1/{10+1/{13+1/{1+1/{13+1/{1+1/{1+1/{5+1/{1+1/{1+1/{1+1/{7+1/{1+1/{1+1/{2+1/{3+1/{2+1/{2+1/{5+1/{1+1/{1+1/{2+1/{4+1/{1+1/{1+1/{1+1/{3+1/{2+1/{5+1/{1+1/{3+1/{46+1/{1+1/{5+1/{1+1/{1+1/{1+1/{1+1/{1+1/{7+1/{951+1/{2+1/{5+1/{5+1/{2+1/{20+1/{14+1/{1+1/{65+1/{3+1/{1+1/{6}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}
3748923572365723865832756328752656235823165278356238756238756237856238756/742389472389478932749832748932657893465783465872365873246587346587342658374568734 = 0+1/{198027369+1/{4+1/{1+1/{6+1/{136+1/{1+1/{4+1/{1+1/{3+1/{21+1/{7+1/{1+1/{1+1/{1+1/{1+1/{4+1/{1+1/{1+1/{1+1/{3+1/{2+1/{3+1/{3+1/{6+1/{2+1/{3+1/{1+1/{2+1/{1+1/{3+1/{1+1/{4+1/{2+1/{1+1/{38+1/{1+1/{29+1/{1+1/{1+1/{1+1/{1+1/{2+1/{4+1/{2+1/{14+1/{1+1/{14+1/{1+1/{1+1/{2+1/{137+1/{1+1/{12+1/{7+1/{1+1/{10+1/{1+1/{5+1/{3+1/{18+1/{1+1/{6+1/{1+1/{2+1/{5+1/{2+1/{1+1/{1306+1/{1+1/{5+1/{5+1/{2+1/{1+1/{1+1/{4+1/{2+1/{13+1/{12+1/{1+1/{5+1/{18+1/{1+1/{1+1/{1+1/{20+1/{8+1/{1+1/{1+1/{1+1/{5+1/{3+1/{20+1/{23+1/{6+1/{3+1/{9+1/{1+1/{1+1/{3+1/{21+1/{6+1/{2+1/{3+1/{1+1/{3+1/{1+1/{15+1/{6+1/{2+1/{1+1/{3+1/{1+1/{2+1/{20+1/{15+1/{1+1/{4+1/{1+1/{1+1/{1+1/{5+1/{46+1/{3+1/{1+1/{2+1/{3+1/{2+1/{4+1/{19}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}
7156217803516290126375612712576239586129561928756283795623789562378915612793856219835623879567321856231578216352357238957238957238915621785689650726521789461239321751107501758291752365012657126357102521307865012786501256015601010120965320715621780351629012637561271257623958612956192875628379562378956237891561279385621983562387956732185623157821635/27589123758192057892137580219375890123758091237589016718297682197689127638496713896713489678134907608932147629837610987618920768591437589127512938758123957129857210957213895721389572139857213895721985712389571230985721389047123980472198347829137589127618927528913571328957129867129471209572198768901237509821375890213758219305789132075 = 259385468934706+1/{5+1/{1+1/{2+1/{3+1/{1+1/{1+1/{1+1/{1+1/{1+1/{1+1/{10+1/{24+1/{3+1/{1+1/{13+1/{3+1/{1+1/{1+1/{1+1/{2+1/{4+1/{1+1/{2+1/{1+1/{1+1/{28+1/{4+1/{2+1/{20+1/{3+1/{1+1/{1+1/{2+1/{4+1/{1+1/{1+1/{2+1/{14+1/{1+1/{2+1/{1+1/{9+1/{1+1/{8+1/{2+1/{26+1/{1+1/{8+1/{1+1/{3+1/{2+1/{1+1/{8+1/{2+1/{4+1/{1+1/{18+1/{1+1/{2+1/{10+1/{2+1/{1+1/{3+1/{1+1/{1+1/{80+1/{2+1/{1+1/{3+1/{1+1/{18+1/{1+1/{1+1/{15+1/{3+1/{1+1/{2+1/{2+1/{1+1/{1+1/{2+1/{1+1/{3+1/{2+1/{10+1/{1+1/{1+1/{1+1/{5+1/{5+1/{109+1/{83+1/{3+1/{20+1/{68+1/{1+1/{1+1/{1+1/{1+1/{10+1/{1+1/{24+1/{160+1/{7+1/{3+1/{83+1/{1+1/{1+1/{3+1/{1+1/{2+1/{1+1/{32+1/{3+1/{6+1/{1+1/{1+1/{6+1/{1+1/{1+1/{6+1/{2+1/{2+1/{1+1/{3+1/{3+1/{2+1/{4+1/{2+1/{2+1/{9+1/{2+1/{1+1/{1+1/{1+1/{1+1/{2+1/{2+1/{2+1/{2+1/{10+1/{2+1/{2+1/{1+1/{8+1/{1+1/{41+1/{4+1/{1+1/{2+1/{2+1/{1+1/{7+1/{1+1/{2+1/{1+1/{7+1/{41+1/{1+1/{2+1/{3+1/{15+1/{3+1/{2+1/{2+1/{7+1/{4+1/{3+1/{1+1/{28+1/{2+1/{4+1/{1+1/{1+1/{22+1/{1+1/{1+1/{3+1/{3+1/{3+1/{1+1/{5+1/{9+1/{4+1/{2+1/{3+1/{4+1/{1+1/{1+1/{2+1/{1+1/{3+1/{1+1/{11+1/{2+1/{2+1/{1+1/{7+1/{7+1/{1+1/{6+1/{1+1/{4+1/{19+1/{1+1/{13+1/{1+1/{3+1/{7+1/{1+1/{1+1/{19+1/{1+1/{7+1/{3+1/{1+1/{1+1/{14+1/{1+1/{3+1/{1+1/{2+1/{4+1/{1+1/{5+1/{1+1/{77+1/{2+1/{1+1/{1+1/{3+1/{1+1/{26+1/{1+1/{1+1/{1+1/{1+1/{80+1/{1+1/{2+1/{35+1/{1+1/{3+1/{1+1/{5+1/{1+1/{1+1/{1+1/{28+1/{2+1/{1+1/{3+1/{5+1/{1+1/{1+1/{1+1/{1+1/{4+1/{2+1/{8+1/{6+1/{4+1/{1+1/{5+1/{7+1/{1+1/{8+1/{1+1/{1+1/{1+1/{1+1/{1+1/{4+1/{1+1/{8+1/{1+1/{1+1/{1+1/{4+1/{4+1/{160+1/{2+1/{1+1/{1+1/{3+1/{1+1/{1+1/{3+1/{2+1/{4+1/{1+1/{1+1/{3+1/{1+1/{1+1/{15+1/{3+1/{1+1/{4+1/{5+1/{7+1/{1+1/{2+1/{4+1/{2+1/{32+1/{1+1/{1+1/{3+1/{1+1/{5+1/{3+1/{1+1/{10+1/{1+1/{3+1/{3+1/{4+1/{6+1/{3+1/{3+1/{2+1/{1+1/{4+1/{2+1/{1+1/{1+1/{3+1/{1+1/{3+1/{4+1/{1+1/{1+1/{54+1/{1+1/{2+1/{1+1/{4+1/{1+1/{6+1/{1+1/{1+1/{3+1/{2+1/{1+1/{2+1/{15+1/{1+1/{2+1/{4+1/{1+1/{2+1/{1+1/{2+1/{1+1/{4+1/{18+1/{601+1/{2+1/{4+1/{1+1/{11+1/{12+1/{1+1/{1+1/{1+1/{4+1/{3+1/{7+1/{1+1/{6+1/{1+1/{2+1/{3+1/{1+1/{5+1/{1+1/{1+1/{2+1/{1+1/{1+1/{2+1/{8+1/{2+1/{10+1/{1+1/{16+1/{1+1/{1+1/{1+1/{2+1/{2+1/{1+1/{4+1/{24+1/{1+1/{2+1/{3+1/{2+1/{1+1/{1+1/{2+1/{7+1/{3+1/{2+1/{2+1/{2+1/{2+1/{11+1/{1+1/{3+1/{1+1/{3+1/{1+1/{15+1/{4+1/{2+1/{2+1/{1+1/{3+1/{3+1/{1+1/{2+1/{1+1/{3+1/{1+1/{1+1/{1+1/{4+1/{1+1/{1+1/{70+1/{1+1/{1+1/{6+1/{4+1/{5+1/{1+1/{1+1/{4+1/{1+1/{5+1/{2+1/{4+1/{4+1/{1+1/{1+1/{20+1/{1+1/{2+1/{7+1/{4+1/{1+1/{88+1/{1+1/{7+1/{1+1/{3+1/{1+1/{4+1/{8+1/{1+1/{1+1/{5+1/{1+1/{2+1/{1+1/{2+1/{3+1/{1+1/{1+1/{1+1/{3+1/{5+1/{2+1/{4+1/{1+1/{4+1/{7+1/{1+1/{1+1/{4+1/{2+1/{9+1/{2+1/{1+1/{2+1/{11+1/{1+1/{3+1/{1+1/{1+1/{1+1/{5+1/{1+1/{2+1/{1+1/{74+1/{1+1/{1+1/{1+1/{1+1/{1+1/{23+1/{1+1/{2+1/{2+1/{1+1/{4+1/{1+1/{10+1/{14+1/{30+1/{7+1/{2+1/{1+1/{1+1/{4+1/{1+1/{1+1/{1+1/{1+1/{1+1/{5+1/{6+1/{1+1/{1+1/{4+1/{4+1/{21+1/{2+1/{2+1/{3+1/{3+1/{1+1/{6+1/{1+1/{4+1/{1+1/{2+1/{1+1/{6+1/{1+1/{1+1/{1+1/{17+1/{1+1/{4+1/{1+1/{1+1/{3+1/{2+1/{2+1/{2+1/{1+1/{1+1/{2+1/{1+1/{3+1/{2+1/{3+1/{1+1/{2+1/{4+1/{2+1/{1+1/{1+1/{6+1/{1+1/{1+1/{1+1/{37+1/{7+1/{6+1/{2+1/{1+1/{7+1/{1+1/{3+1/{1+1/{2+1/{1+1/{1+1/{4+1/{1+1/{43+1/{1+1/{2+1/{1+1/{6+1/{23+1/{1+1/{5+1/{4+1/{5+1/{2+1/{5+1/{2+1/{4+1/{1+1/{17+1/{4+1/{2+1/{10+1/{1+1/{1+1/{4+1/{1+1/{2+1/{1+1/{3+1/{1+1/{1+1/{1+1/{18+1/{1+1/{1+1/{2+1/{1+1/{3+1/{25+1/{11+1/{1+1/{6+1/{5+1/{1+1/{12+1/{2+1/{1+1/{1+1/{1+1/{1+1/{13+1/{4+1/{2+1/{2+1/{5+1/{1+1/{3+1/{2+1/{2+1/{5+1/{1+1/{3+1/{3+1/{2+1/{6+1/{1+1/{1+1/{2+1/{9+1/{72+1/{13+1/{2+1/{4+1/{2+1/{1+1/{30+1/{1+1/{50+1/{7+1/{1+1/{189+1/{12+1/{3+1/{1+1/{4+1/{1+1/{2+1/{3+1/{1+1/{1+1/{1+1/{1+1/{2+1/{10+1/{13+1/{1+1/{3}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}
0/1 = 0
0/3278 = 0
-214/2347826847632478236 = -1+1/{1+1/{10971153493609710+1/{2+1/{1+1/{1+1/{1+1/{1+1/{3+1/{2}}}}}}}}}
19/1718 = 0+1/{90+1/{2+1/{2+1/{1+1/{2}}}}}
-23578/-273589 = 0+1/{11+1/{1+1/{1+1/{1+1/{1+1/{10+1/{1+1/{1+1/{1+1/{1+1/{41+1/{2}}}}}}}}}}}}
-123432521751/-23521512 = 5247+1/{1+1/{1+1/{1+1/{4+1/{4+1/{5+1/{1+1/{1+1/{2+1/{1+1/{1+1/{2+1/{9+1/{5+1/{2}}}}}}}}}}}}}}}
